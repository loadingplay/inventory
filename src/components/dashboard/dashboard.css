.dashboard_wrapper {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: auto 1fr;
  padding: 20px;
}

.dashboard_title {
  font-size: 18px;
  font-weight: 700;
  text-transform: capitalize;
}

.dashboard_elements {
  display: grid;
  grid-template-columns: 200px 1fr;
  grid-template-rows: auto;
  margin-top: 20px;
  grid-gap: 20px;
}

.subtitle {
  font-weight: 700;
}
