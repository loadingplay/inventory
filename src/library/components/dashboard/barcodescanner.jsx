import React from 'react';

export function withBarcodeScanner(WrappedComponent) {
  return class BarCodeScanner extends React.Component {
    constructor(props) {
      super(props);

      this.timer = null;
      this.wrapped = null;
      this.input_string = "";

      this.initListener();
    }

    initListener = () => {
      if (typeof window !== `undefined`)
        document.addEventListener('keyup', this.handleKeyUp);
    }

    handleKeyUp = (e) => {
      const timeoutCallback = () => {
        if (this.input_string.length <= 3) {
          this.input_string = "";
          return;
        }

        this.handleScannerRead(this.input_string);
        this.input_string = "";
      }

      // border cases
      if (e.key === "Enter") timeoutCallback();
      if (this.timer) clearTimeout(this.timer);
      if (e.key.length <= 1) this.input_string += e.key;

      this.timer = setTimeout(timeoutCallback, 100);
    }

    handleScannerRead = (input_string) => {
      if (!this.wrapped || !this.wrapped.handleScannerRead) {
        console.log("implement handleScannerRead method");
        return;
      }

      this.wrapped.handleScannerRead(input_string);
    }

    render = () => {
      return <WrappedComponent {...this.props} ref={e => this.wrapped = e} />;
    }
  }
}
