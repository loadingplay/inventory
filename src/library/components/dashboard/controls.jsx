import React, { Component } from 'react';
import PropTypes from "prop-types";

import {
  LogoutButton,
  PickerClearButton,
  SearchInput,
  GenerateOrderButton,
  PrintButton
} from '../../components';


export class DashboardControls extends Component {

  static propTypes = {
    selectedCellarID: PropTypes.number,
    accessToken: PropTypes.string,
    onSearchTermChange: PropTypes.func,
    onPickerCleared: PropTypes.func
  }

  render = () => {
    return (
      <React.Fragment>
        <div className="col-6" >
          <SearchInput
            onSearch={this.props.onSearchTermChange}
            onTypeChange={this.props.onSearchTypeChange}
          ></SearchInput>
        </div>
        <div className="col-6 text-right">
          <PickerClearButton
            selected_cellar_id={this.props.selectedCellarID}
            onPickerClear={this.props.onPickerCleared}
          ></PickerClearButton>
          <PrintButton></PrintButton>
          <GenerateOrderButton
            selectedCellar={this.props.selectedCellarID}
            accessToken={this.props.accessToken}
            onOrderGenerated={this.props.onPickerCleared}
          ></GenerateOrderButton>
          <LogoutButton></LogoutButton>
        </div>
      </React.Fragment>
    );
  }
}
