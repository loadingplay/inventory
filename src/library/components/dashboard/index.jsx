export * from './layout';
export * from './controls';
export * from './dashboard';
export * from './printbutton';
