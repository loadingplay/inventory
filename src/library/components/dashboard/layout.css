body {
    padding-top: 4.5rem;
}

.toast-container {
    width: 100%;
    position: absolute;
    bottom: 150px;
    padding: 10px;
}

.toast-content {
    background: #333;
    color: #ccc;
    border-radius: 5px;
    padding: 10px;
    width: fit-content;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    font-size: 25px;
}
