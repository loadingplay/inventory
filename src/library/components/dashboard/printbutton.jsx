import React from "react";


export class PrintButton extends React.Component {

  render = () => {
    return (
      <a className="btn btn-warning ml-1" role="button" href="/print" target="_blank" >
        Imprimir
      </a>
    );
  }
}
