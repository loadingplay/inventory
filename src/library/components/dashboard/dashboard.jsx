import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';

import { Stock, Stores, DashboardLayout, DashboardControls } from '../../components';
import { withBarcodeScanner } from "./barcodescanner";


export class Dashboard extends React.Component {

  static propTypes = {
    onInventoryRequest: PropTypes.func,
    accessToken: PropTypes.string,
    extraProducts: PropTypes.array,
    products: PropTypes.array,
    isLoading: PropTypes.bool,
    hasError: PropTypes.bool,
    stockErrorMessage: PropTypes.string,
    pageCount: PropTypes.number,
    onScannerRead: PropTypes.func,
    onPickerClear: PropTypes.func,
    loadKey: PropTypes.number,
    isScannerLoading: PropTypes.number
  };

  constructor(props) {
    super(props);
    this.state = {
      selected_cellar_id: 0,
      selected_cellar_name: "",
      current_page: 0,
      search_term: "",
      search_type: 'search'
    };
  }

  reloadInventory = async () => {
    // reset status
    await this.props.onInventoryRequest(
      this.state.hq_cellar_id,
      this.state.selected_cellar_id,
      this.state.current_page + 1,
      this.state.search_term,
      this.state.search_type
    );
  }

  handleSelectedCellar = (cellar_id, cellar_name) => {
    if (this.state.selected_cellar_id === cellar_id) return;
    this.setState({
      selected_cellar_id: cellar_id,
      selected_cellar_name: cellar_name
    }, () => {
      this.reloadInventory();
    });
  }

  handlePageClick = (e) => {
    this.setState({
      current_page: e.selected
    }, () => {
      this.reloadInventory();
    });
  }

  handleHQCellar = (hq_cellar_id) => {
    this.setState({
      hq_cellar_id: hq_cellar_id
    });
  }

  handleSearch = (search_term) => {
    if (this.state.search_term === search_term) return;
    this.setState({
      search_term
    }, () => {
      this.reloadInventory();
    });
  }

  handleSearchType = (type) => {
    this.setState({
      search_type: type,
      search_term: ''
    }, () => {
      this.reloadInventory();
    });
  }

  handleScannerRead = (input_string) => {
    if (this.state.search_type === "barcode")
      this.props.onScannerRead(this.state.hq_cellar_id, this.state.selected_cellar_id, input_string);
  }

  render() {
    return (
      <DashboardLayout
        toastMessage={this.props.isScannerLoading ? `Procesando cola de scanner quedan ${this.props.isScannerLoading} ...`:''}
        title="ReposiciÃ³n de inventario."
        menuTitle="1. Seleccion de tienda"
        menu={
          <Stores
            accessToken={this.props.accessToken}
            cellarSelected={this.handleSelectedCellar}
            onHQCellarLoaded={this.handleHQCellar}
          />
        }

        tableTitle={
          "2. Productos" + (this.state.selected_cellar_name !== "" ?
          ` | usted esta reponiendo para la bodega ${this.state.selected_cellar_name}`:
          "")
        }
        table={
          <Stock
            loadKey={this.props.loadKey}
            selectedCellar={this.state.selected_cellar_id}
            extraProducts={this.props.extraProducts}
            products={this.props.products}
            isLoading={this.props.isLoading}
            hasError={this.props.hasError}
            errorMessage={this.props.stockErrorMessage}
          />
        }
        tableControls={
          <DashboardControls
            selectedCellarID={this.state.selected_cellar_id}
            accessToken={this.props.accessToken}

            onSearchTermChange={this.handleSearch}
            onSearchTypeChange={this.handleSearchType}
            onPickerCleared={this.props.onPickerClear}
          ></DashboardControls>
        }
        tableDownControls={
          <ReactPaginate
            previousLabel={'anterior'}
            nextLabel={'siguiente'}
            breakLabel={'...'}
            pageCount={this.props.pageCount}
            forcePage={this.state.current_page}
            marginPagesDisplayed={2}
            pageRangeDisplayed={5}
            onPageChange={this.handlePageClick}
            containerClassName={'pagination'}
            previousClassName="page-item"
            previousLinkClassName="page-link"
            nextClassName="page-item"
            nextLinkClassName="page-link"
            pageClassName="page-item"
            pageLinkClassName="page-link"
            breakClassName="page-item"
            breakLinkClassName="page-link"
            activeClassName={'active'}
          />
        }
      />
    );
  }
}

export const DashboardWithScanner = withBarcodeScanner(Dashboard);
