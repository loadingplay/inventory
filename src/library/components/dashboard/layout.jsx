import React, { Component } from 'react';
import PropTypes from "prop-types";

import { LogoutButton } from '../../components';

export class DashboardLayout extends Component {

  static propTypes = {
    title: PropTypes.string,

    menuTitle: PropTypes.string,
    menu: PropTypes.object,

    tableTitle: PropTypes.string,
    table: PropTypes.object,
    tableControls: PropTypes.object
  }

  render = () => {
    return (
      <div className="container dashboard_wrapper">
        <div className="dashboard_elements row">
          <h3 className="dashboard_title col-12">
            {this.props.title}
          </h3>

          {/* Stores */}
          <div className="col-2" >
            <div className="subtitle">{this.props.menuTitle}</div>
            {this.props.menu}
          </div>

          {/* Inventory table */}
          <div className="col-10 row" >
            <div className="subtitle col-6">
              {this.props.tableTitle}
            </div>
            <div className="col-6">
              <LogoutButton></LogoutButton>
            </div>
            {this.props.table}

            {/* Controls */}
            <section className="controls col-12 row">
              {this.props.tableControls}
            </section>
          </div>
        </div>
      </div>
    )
  }
}
