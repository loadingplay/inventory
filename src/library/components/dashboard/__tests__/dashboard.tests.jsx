import React from 'react';
import { shallow } from "enzyme";

import { Dashboard } from "../dashboard";

describe("Dashboard", () => {
  test("it should render properly", () => {
    const wrapper = shallow(<Dashboard />);
    expect(wrapper).toMatchSnapshot("detault");
  });

  test("it should handle scanner reads", () => {

    const onScannerRead = jest.fn();
    const wrapper = shallow(<Dashboard onScannerRead={onScannerRead} />);

    wrapper.instance().setState({ search_type: "search" });
    wrapper.instance().handleScannerRead("test");
    expect(onScannerRead).toHaveBeenCalledTimes(0);

    wrapper.instance().setState({ search_type: "barcode" });
    wrapper.instance().handleScannerRead("test");
    expect(onScannerRead).toHaveBeenCalledTimes(1);
  });
});
