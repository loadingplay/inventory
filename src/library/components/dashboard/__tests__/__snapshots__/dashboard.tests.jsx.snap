// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard it should render properly: detault 1`] = `
<DashboardLayout
  menu={
    <Stores
      cellarSelected={[Function]}
      onHQCellarLoaded={[Function]}
    />
  }
  menuTitle="1. Seleccion de tienda"
  table={
    <Stock
      selectedCellar={0}
    />
  }
  tableControls={
    <DashboardControls
      onSearchTermChange={[Function]}
      onSearchTypeChange={[Function]}
      selectedCellarID={0}
    />
  }
  tableDownControls={
    <o
      activeClassName="active"
      breakClassName="page-item"
      breakLabel="..."
      breakLinkClassName="page-link"
      containerClassName="pagination"
      disableInitialCallback={false}
      disabledClassName="disabled"
      eventListener="onClick"
      forcePage={0}
      marginPagesDisplayed={2}
      nextAriaLabel="Next page"
      nextClassName="page-item"
      nextLabel="siguiente"
      nextLinkClassName="page-link"
      onPageChange={[Function]}
      pageClassName="page-item"
      pageCount={10}
      pageLinkClassName="page-link"
      pageRangeDisplayed={5}
      previousAriaLabel="Previous page"
      previousClassName="page-item"
      previousLabel="anterior"
      previousLinkClassName="page-link"
    />
  }
  tableTitle="2. Productos"
  title="ReposiciÃ³n de inventario."
  toastMessage=""
/>
`;
